lib/format_parser.rb

```diff
@@ -8,21 +8,34 @@ module FormatParser
   require_relative 'remote_io'
   require_relative 'io_constraint'
   require_relative 'care'
-  require_relative 'parsers/dsl'
 
   PARSER_MUX = Mutex.new
 
-  def self.register_parser_constructor(object_responding_to_new)
+  def self.register_parser(callable_or_responding_to_new, formats:, natures:)
+    parser_provided_formats = Array(formats)
+    parser_provided_natures = Array(natures)
     PARSER_MUX.synchronize do
-      @parsers ||= []
-      @parsers << object_responding_to_new
-      # Gathering natures and formats from parsers. An instance has to be created.
-      parser = object_responding_to_new.new
-      @natures ||= Set.new
-      # NOTE: merge method for sets modify the instance.
-      @natures.merge(parser.natures)
-      @formats ||= Set.new
-      @formats.merge(parser.formats)
+      @parsers ||= Set.new
+      @parsers << callable_or_responding_to_new
+      @natures_per_parser ||= {}
+      parser_provided_natures.each do |provided_nature|
+        @natures_per_parser[provided_nature] ||= Set.new
```

Comment:
Should be this call parsers_per_nature? After all, it's an inverted index for parsers given a nature. Same goes for format.

Reply:
Yep.